man/iperender.1
.\"                                      Hey, EMACS: -*- nroff -*-
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.\" TeX users may be more comfortable with the \fB<whatever>\fP and
.\" \fI<whatever>\fP escape sequences to invode bold face and italics, 
.\" respectively.
.TH IPERENDER 1 "June 27, 2009"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
iperender \- Export Ipe document as bitmap or SVG 
.SH SYNOPSIS
.B iperender
( -png | -eps | -pdf | -svg ) 
[ -page \fIpage\fP ] 
[ -view \fIview\fP ]
[ -resolution \fIdpi\fP ]
[ -transparent ]
[ -nocrop ] 
\fIinput-file\fP \fIoutput-file\fP

.SH DESCRIPTION
.PP
\fBiperender\fP exports an Ipe document as a PNG bitmap, as a
figure in Encapsulated Postscript (EPS), in PDF, or in SVG format.
Options are:
.TP
\fB-png\fP
export as a PNG image
.TP
\fB-eps\fP
export as an EPS figure
.TP
\fB-pdf\fP
export as a PDF figure
.TP
\fB-svg\fP
export as an SVG figure
.TP
\fB-page\fP \fIpage\fP
export this page from a multipage document.
.TP
\fB-view\fP \fIview\fP
export this view from a page with multiple views.
.TP
\fB-resolution\fP \fIdpi\fP
set resolution for exporting to PNG.
.TP
\fB-transparent\fP
make background transparent when exporting to PNG.
.TP
\fB-nocrop\fP
do not crop the page to the bounding box of the objects.

.SH ENVIRONMENT VARIABLES

\fBiperender\fP respects the following environment variables:

.TP
\fBIPELATEXDIR\fP
the directory where \fBipetoipe\fP runs Pdflatex.
.TP
\fBIPELATEXPATH\fP
the directory containing pdflatex and xelatex.  If not set, they are
searched for on the path.
.TP
\fBIPEDEBUG\fP
set this to 1 for debugging output from \fBiperender\fP.

.SH AUTHOR
Otfried Cheong

.SH REPORTING BUGS
.ad l
Please report bugs using Ipe bugzilla at
.I "http://ipe.otfried.org/bugzilla.html"

.SH SEE ALSO
.ad l
Ipe is documented fully in
.IR "The Ipe Manual" ,
which is available online at
.I "http://ipe.otfried.org/manual/manual.html" .


doc/news.txt
----------------------------------------------------------------------

Ipe 7.2.28

 * Add "sequential text" option.

 * Fix centering of title in presentation stylesheet (bug #464).

----------------------------------------------------------------------

Ipe 7.2.27 (2023-05-05)

 * On Linux, migrate to Qt6 (#457).  Thanks Henri!

 * Change behaviour of transform tools to never switch the current
   selection (#459).

 * Action to deselect everything on current page.

 * Align center/H center/V center now aligns to the page if
   only one object selected (#449).

 * Fix absolute autosave paths (bug #433).

 * Add /opt/homebrew/bin to path on MacOS (#440).

 * Fix missing 'stroke-opacity' and 'opacity' attributes in
   'ipe.dtd' (#67).

 * Fix crash when 'recent_files_lua' ends up empty (bug #455).

 * Fix missing page numbers when there is no text (bug #450).

 * Don't lose group's link action during group edit (bug #456).

----------------------------------------------------------------------

Ipe 7.2.26 (2022-07-16)

 * Allow to store autosave file in same directory as document
   itself.  This is now also the default. (#426)

 * Circular graph edges no longer get broken when moving graph
   nodes (bug #424).

 * Make it possible to remap left_command on MacOS (bug #431).

 * Use PKG_CONFIG environment variable (#429).

----------------------------------------------------------------------

Ipe 7.2.25 (2022-06-19) Denmark Edition

 * Add spell-checking to text entry on Linux.

 * Change presentation style sheet to have 16:9 aspect ratio (#385).

 * Many more configuration options in prefs.canvas_style (#400 &
   #406).

 * Remove box around links in PDF output (#397).

 * Add 'script' text size (#410).

 * Use most-recently used directory in "Save" and "Save As" (#394).

 * Maintain a single custom string on each Ipe object. It's not
   used by Ipe itself, but can be used by external software or
   ipelets. (#409)

 * Add page transition effects from PDF 1.5 (Push, Cover, Uncover,
   Fade).

 * Adjust highlighting text colors in dark theme.

 * Fix crash when compiling with -D_GLIBCXX_ASSERTIONS (#387).

 * Try to simply use UTF-8 in 'runlatex.bat'.  This should work
   at least on Windows-10 (bug #338, second case).

 * Upgrade Freetype in Windows package (fixes bug #411).

 * Show error message when a style sheet mentioned in prefs.styles
   cannot be read (bug #405).

 * Reuse PDF text resources for identical text objects (#412).

 * Fix losing primary selection when deselecting using rectangle drag
   (bug #423).

----------------------------------------------------------------------

Ipe 7.2.24 (2021-04-07)

 * Fix a crash on MacOS 11 El Sur by applying a patch to cairo
   (bug #370).

 * Switch from CAIRO_FILTER_FAST to CAIRO_FILTER_GOOD for decent
   display (essential when filming from the canvas).

 * Make "Select by type/attribute" ipelet work when selecting on an
   absolute color (#380).

 * Fix semi-transparent bitmaps as well as bitmaps with color keys not
   working in Ipepresenter (bug #9).

 * Fix exporting to EPS when part of the drawing lies outside the
   paper (bug #369).

 * Remember online latex service URL in a different file to avoid
   problems when using old and new Ipe versions.

 * Improve calling of external editor on Qt, so that keyboard focus
   is given to the external editor.

 * Raise Ipe window after returning from external editor on Qt, in
   case the editor program is still running.

 * New preference 'freeze_in_latex' to let the Ipe UI freeze during a
   Latex run.

 * New shell script 'iperender-par' that allows to run several
   instances of iperender in parallel (e.g. as part of a parallel
   make) (#375).

 * Add Lua API to access (Latex) dimensions of text object.

 * Silently ignore non-integer grid sizes (#372).

 * On MacOS 11, return to the classic toolbar underneath the title
   bar, because the new unified look leaves most tool buttons off the
   screen (bug #376).

----------------------------------------------------------------------

Ipe 7.2.23 (2020-12-24)

 * Shortcut for pretty display.

 * A maximum on the number of recent files remembered.

 * Improve scaling of toolbar buttons on El Sur, and add labels for
   the toolbar overflow menu.
 
 * Remove sources of libspiro from Ipe releases.

 * Remove doxygen documentation from Ipe release, run doxygen as part
   of the build process.

----------------------------------------------------------------------

Ipe 7.2.22 (2020-12-20)

 * Recent file menu (#322).

 * Update PDF resources on main thread to avoid race condition (bugs
   #354, #359).

 * Can exclude pages from PDF output by adding a layer "NOPDF" to the
   page. The layer does not need to be visible. (#362).  

 * Use a better bitmap checksum that is less likely to cause bitmaps
   to be reordered.  Sort mappings of symbolic attributes to make them
   deterministic (#356).

 * Display Type3 fonts with a unique representation to make it easy to
   diagnose the problem.

----------------------------------------------------------------------

Ipe 7.2.21 (2020-10-26)

 * Fix hanging with message "Running Latex" on Windows (bug #338).

 * Ability to hide the properties panel/layer list from the UI on
   macOS, for use on Zoom etc (#350).
 
 * Remove NoFocus policy from selectors in Qt.  This used to be
   necessary to keep shortcuts working when the selector receives
   focus, but that's no longer necessary (tested in Qt 5.9).
   Fixes selectors not working on Qt 5.15 (bug #341).

 * Fix incorrect update of grid size and angle size selector when
   style sheet is changed (bug #344).

 * Fix crash in 'iperender -pdf' when the document uses
   \includegraphics on a jpg file (bug #342).

 * Fix snapping to midpoint of closing segment of closed paths (bug
   #346).

 * Fix Fifi remaining invisible after switching back from ink mode
   (bug #347).

 * Escape key now closes page/view selectors and sorters on Windows
   (bug #349).


src/iperender/iperender.cpp
// --------------------------------------------------------------------
// iperender
// --------------------------------------------------------------------
/*

    This file is part of the extensible drawing editor Ipe.
    Copyright (c) 1993-2023 Otfried Cheong

    Ipe is free software; you can redistribute it and/or modify it
    under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 3 of the License, or
    (at your option) any later version.

    As a special exception, you have permission to link Ipe with the
    CGAL library and distribute executables, as long as you follow the
    requirements of the Gnu General Public License in regard to all of
    the software in the executable aside from CGAL.

    Ipe is distributed in the hope that it will be useful, but WITHOUT
    ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
    or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public
    License for more details.

    You should have received a copy of the GNU General Public License
    along with Ipe; if not, you can find it at
    "http://www.gnu.org/copyleft/gpl.html", or write to the Free
    Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

*/

#include "ipedoc.h"
#include "ipethumbs.h"

#include <cstdio>
#include <cstdlib>
#include <cstring>

using ipe::Document;
using ipe::Page;
using ipe::Thumbnail;

// --------------------------------------------------------------------

static int renderPage(Thumbnail::TargetFormat fm,
		      const char *src, const char *dst,
		      const char *pageSpec, const char *viewSpec, double zoom,
		      bool transparent, bool nocrop)
{
  Document *doc = Document::loadWithErrorReport(src);

  if (!doc)
    return 1;

  int pageIdx = pageSpec ? doc->findPage(pageSpec) : 0;
  if (pageIdx < 0) {
    fprintf(stderr, "Incorrect -page specification.\n");
    delete doc;
    return 1;
  }

  const Page *page = doc->page(pageIdx);

  int viewIdx = viewSpec ? page->findView(viewSpec) : 0;
  if (viewIdx < 0) {
    fprintf(stderr, "Incorrect -view specification.\n");
    delete doc;
    return 1;
  }

  if (doc->runLatex(src)) {
    delete doc;
    return 1;
  }

  Thumbnail tn(doc, 0);
  tn.setTransparent(transparent);
  tn.setNoCrop(nocrop);
  if (!tn.saveRender(fm, dst, page, viewIdx, zoom))
    fprintf(stderr, "Failure to render page.\n");
  delete doc;
  return 0;
}

// --------------------------------------------------------------------

static void usage()
{
  fprintf(stderr, "Usage: iperender [ -png ");
#ifdef CAIRO_HAS_PS_SURFACE
  fprintf(stderr, "| -eps ");
#endif
#ifdef CAIRO_HAS_PDF_SURFACE
  fprintf(stderr, "| -pdf ");
#endif
#ifdef CAIRO_HAS_SVG_SURFACE
  fprintf(stderr, "| -svg ");
#endif
  fprintf(stderr, "] "
	  "[ -page <page> ] [ -view <view> ] [ -resolution <dpi> ] "
	  "[ -transparent ] [ -nocrop ] "
	  "infile outfile\n"
	  "Iperender saves a single page of the Ipe document in some formats.\n"
	  " -page       : page to save (default 1).\n"
	  " -view       : view to save (default 1).\n"
	  " -resolution : resolution for png format (default 72.0 ppi).\n"
	  " -transparent: use transparent background in png format.\n"
	  " -nocrop     : do not crop page.\n"
	  "<page> can be a page number or a page name.\n"
	  );
  exit(1);
}

int main(int argc, char *argv[])
{
  ipe::Platform::initLib(ipe::IPELIB_VERSION);

  // ensure at least three arguments (handles -help as well :-)
  if (argc < 4)
    usage();

  Thumbnail::TargetFormat fm = Thumbnail::EPNG;
  if (!strcmp(argv[1], "-png"))
    fm = Thumbnail::EPNG;
#ifdef CAIRO_HAS_PS_SURFACE
  else if (!strcmp(argv[1], "-eps"))
    fm = Thumbnail::EPS;
#endif
#ifdef CAIRO_HAS_PDF_SURFACE
  else if (!strcmp(argv[1], "-pdf"))
    fm = Thumbnail::EPDF;
#endif
#ifdef CAIRO_HAS_SVG_SURFACE
  else if (!strcmp(argv[1], "-svg"))
    fm = Thumbnail::ESVG;
#endif
  else
    usage();

  const char *page = nullptr;
  const char *view = nullptr;
  double dpi = 72.0;
  bool transparent = false;
  bool nocrop = false;

  int i = 2;
  while (i < argc - 2) {
    if (!strcmp(argv[i], "-page")) {
      if (i + 1 == argc)
	usage();
      page = argv[i+1];
      i += 2;
    } else if (!strcmp(argv[i], "-view")) {
      if (i + 1 == argc)
	usage();
      view = argv[i+1];
      i += 2;
    } else if (!strcmp(argv[i], "-resolution")) {
      if (i + 1 == argc)
	usage();
      dpi = ipe::Lex(ipe::String(argv[i+1])).getDouble();
      i += 2;
    } else if (!strcmp(argv[i], "-transparent")) {
      transparent = true;
      ++i;
    } else if (!strcmp(argv[i], "-nocrop")) {
      nocrop = true;
      ++i;
    } else
      usage();
  }

  // remaining arguments must be two filenames
  const char *src = argv[i];
  const char *dst = argv[i+1];

  return renderPage(fm, src, dst, page, view, dpi / 72.0,
		    transparent, nocrop);
}

// --------------------------------------------------------------------
